### PR TITLE
BE-74-Document-Diff

### PR DESCRIPTION
**Pull request by Scr0ll**

### CODE REVIEW

**Review by MCFreddie777 — CHANGES_REQUESTED** [review]
Please resolve your commits. It seems like there are changes that are already commited.

File: antiplag/tasks.py [hunk 1]
```diff
@@ -21,7 +21,6 @@ def process_documents(submission_id):
     submission.save()
 
     documents = submission.documents.all()
-
```
Comment (MCFreddie777) [inline antiplag/tasks.py:24]: Have you removed the empty lines on purpose or did `black` formatter do this?

File: antiplag/views.py [hunk 2]
```diff
@@ -124,3 +123,55 @@ def get(self, request, id):
...
+
+class DocumentDiff(APIView):
+    def get(self, request, first_id, second_id):
+        first_document = get_object_or_404(Document, pk=first_id)
+        second_document = None
+        intervals = []
+
+        if first_document.submission.status == Submission.SubmissionStatus.PROCESSED:
+
+            for doc in first_document.result.matched_docs:
```
Comment (MCFreddie777) [inline antiplag/views.py:136]:
```
second_document = next(doc for doc in first_document.result.matched_docs if doc["elastic_id"] == second_id)
```

The reason is that the next will end after finding first match, therefore it reduced the complexity.

File: antiplag/views.py [hunk 3]
```diff
@@ -124,3 +123,55 @@ def get(self, request, id):
...
+                if doc["elastic_id"] == second_id:
+                    second_document = doc
+
+            intervals = second_document["intervals"]
+
+            if not second_document:
```
Comment (MCFreddie777) [inline antiplag/views.py:142]: I think this should be above `intervals`, shouldn't it?

File: antiplag/views.py [hunk 4]
```diff
@@ -124,3 +123,55 @@ def get(self, request, id):
...
+                return Response(status=status.HTTP_404_NOT_FOUND)
+
+            matches = []
+            second_text = preprocess_text(second_document["text"],
```
Comment (MCFreddie777) [inline antiplag/views.py:146]: Should we preprocess the text here? I think this operation is super-expensive for performance -therefore we do it in celery.

Doing this on endpoint blocks server for other users.

Comment (MCFreddie777) [reply antiplag/views.py:146]: Cannot we find the matches it on raw text instead? When comparing , the text from elastic wasn't preprocessed either, was it?

**Review by MCFreddie777 — CHANGES_REQUESTED** [review]

File: antiplag/tasks.py (hunk 1 above)
Comment (Scr0ll) [reply antiplag/tasks.py]: it was my bad. fixed

File: antiplag/views.py (hunk 2 above)
Comment (Scr0ll) [reply antiplag/views.py]: done

File: antiplag/views.py (hunk 3 above)
Comment (Scr0ll) [reply antiplag/views.py:137]: fixed

File: antiplag/views.py (hunk 4 above)
Comment (Scr0ll) [reply antiplag/views.py]: refactored code. intervals are now saved directly in greedy string tiling algorithm.

**Review by MCFreddie777 — APPROVED** [review]
LGTM. There's one unused import in views, but it can be fixed somewhen later.

**Pull request merged**